tools/pubsys/src/aws/ssm/ssm.rs

```diff
@@ -135,6 +135,84 @@ where
     Ok(parameters)
 }
 
+/// Fetches all SSM parameters under a given prefix using the given clients
+pub(crate) async fn get_parameters_by_prefix<'a>(
+    clients: &'a HashMap<Region, SsmClient>,
+    ssm_prefix: &str,
+) -> HashMap<&'a Region, Result<SsmParameters>> {
+    // Build requests for parameters; we have to request with a regional client so we split them by
+    // region
+    let mut requests = Vec::with_capacity(clients.len());
+    for region in clients.keys() {
+        trace!("Requesting parameters in {}", region);
+        let ssm_client: &SsmClient = &clients[region];
+        let get_future = get_parameters_by_prefix_in_region(region, ssm_client, ssm_prefix);
+
+        requests.push(join(ready(region), get_future));
+    }
+
+    // Send requests in parallel and wait for responses, collecting results into a list.
+    requests
+        .into_iter()
+        .collect::<FuturesUnordered<_>>()
+        .collect()
+        .await
+}
+
+/// Fetches all SSM parameters under a given prefix in a single region
+pub(crate) async fn get_parameters_by_prefix_in_region(
+    region: &Region,
+    client: &SsmClient,
+    ssm_prefix: &str,
+) -> Result<SsmParameters> {
+    info!("Retrieving SSM parameters in {}", region.to_string());
+    let mut parameters = HashMap::new();
+
+    // Send the request
+    let mut get_future = client
+        .get_parameters_by_path()
+        .path(ssm_prefix)
+        .recursive(true)
+        .into_paginator()
+        .send();
+
+    // Iterate over the retrieved parameters
+    while let Some(page) = get_future.next().await {
+        for parameter in page
+            .context(error::GetParametersByPathSnafu {
+                path: ssm_prefix,
+                region: region.to_string(),
+            })?
+            .parameters()
+            .unwrap_or_default()
+        {
+            // Insert a new key-value pair into the map, with the key containing region and parameter name
+            // and the value containing the parameter value
+            parameters.insert(
+                SsmKey::new(
+                    region.to_owned(),
+                    parameter
+                        .name()
+                        .ok_or(error::Error::MissingField {
+                            region: region.to_string(),
+                            field: "name".to_string(),
+                        })?
+                        .to_owned(),
+                ),
+                parameter
+                    .value()
+                    .ok_or(error::Error::MissingField {
+                        region: region.to_string(),
+                        field: "value".to_string(),
+                    })?
```

Comment:
Is it possible that a "null" value is something we could encounter and would want to report up through the results instead of erroring out here?

Reply:
Based on the information in https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-parameter-store.html, I don't think it's possible to have a parameter without the field "value", so this error will essentially never happen. If the "value" field contains a "null" value (which I assume means empty string), it'll be propagated in the results as usual.